spectrum-network/src/peer_manager.rs

```diff
@@ -292,31 +352,75 @@ impl<S: Unpin + PeersState> Stream for PeerManager<S> {
                 match notif {
-                    InNotification::IncomingConnection(pid, index) => {
-                        self.on_incoming_connection(pid, index)
-                    }
-                    InNotification::ConnectionLost(pid, reason) => {
-                        self.on_connection_lost(pid, reason)
-                    }
+                    PeerManagerIn::Notification(notification) => match notification {
+                        PeerEvent::IncomingConnection(pid, conn_id) => {
+                            self.on_incoming_connection(pid, conn_id)
+                        }
+                        PeerEvent::ConnectionEstablished(pid, conn_id) => {
+                            self.on_incoming_connection(pid, conn_id)
+                        }
+                        PeerEvent::ConnectionLost(pid, reason) => self.on_connection_lost(pid, reason),
+                        PeerEvent::ForceEnabled(pid, protocol_id) => {
+                            self.on_force_enabled(pid, protocol_id);
+                        }
+                    },
+                    PeerManagerIn::Request(req) => match req {
+                        PeerManagerRequest::AddPeer(pid) => self.on_add_peer(pid),
+                        PeerManagerRequest::ReportPeer(pid, adjustment) => {
+                            self.on_report_peer(pid, adjustment)
+                        }
+                        PeerManagerRequest::AddReservedPeer(pid) => self.on_add_reserved_peer(pid),
+                        PeerManagerRequest::GetPeerReputation(pid, resp) => {
+                            self.on_get_peer_reputation(pid, resp)
+                        }
+                        PeerManagerRequest::SetReservedPeers(peers) => self.on_set_reserved_peers(peers),
+                        PeerManagerRequest::SetProtocols(pid, protocols) => {
+                            self.on_set_peer_protocols(pid, protocols)
+                        }
+                    },
                 }
             }
 
-            if let Poll::Ready(Some(req)) = Stream::poll_next(Pin::new(&mut self.requests_recv), cx)
-            {
-                match req {
-                    InRequest::AddPeer(pid) => self.on_add_peer(pid),
-                    InRequest::ReportPeer(pid, adjustment) => self.on_report_peer(pid, adjustment),
-                    InRequest::AddReservedPeer(pid) => self.on_add_reserved_peer(pid),
-                    InRequest::GetPeerReputation(pid, resp) => {
-                        self.on_get_peer_reputation(pid, resp)
+            // Allocate protocol substreams according to defined policies.
+            for (prot, policy) in self.conf.protocols_allocation.clone().iter() {
+                if let Some(enabled_peers) = self.state.get_enabled_peers(prot) {
+                    let cond = match policy {
+                        ProtocolAllocationPolicy::Bounded(max_conn_percent) => {
```

Comment:
```suggestion
                        ProtocolAllocationPolicy::Bounded(max_conn_percent) => {
```
Maybe incapsulate it in "Impl ProtocolAllocationPolicy"